vicky/src/lib/database/entities/task.rs

```diff
@@ -0,0 +1,415 @@
+    };
+    use std::collections::HashMap;
+    use std::fmt::Display;
+    use uuid::Uuid;
+    // these here are evil >:(
+    use crate::database::schema::locks;
+    use crate::database::schema::tasks;
+    use itertools::Itertools;
+    use rocket_sync_db_pools::database;
+
+    #[derive(Insertable, Queryable, AsChangeset, Debug)]
+    #[diesel(table_name = tasks)]
+    struct DbTask {
+        pub id: Uuid,
+        pub display_name: String,
+        pub status: String,
+        pub features: String,
+        pub flake_ref_uri: String,
+        pub flake_ref_args: String,
+    }
+
+    impl Display for TaskStatus {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            let str = match self {
+                TaskStatus::NEW => "NEW",
+                TaskStatus::RUNNING => "RUNNING",
+                TaskStatus::FINISHED(r) => match r {
+                    TaskResult::SUCCESS => "FINISHED::SUCCESS",
+                    TaskResult::ERROR => "FINISHED::ERROR",
+                },
+            };
+            write!(f, "{}", str)
+        }
+    }
+
+    impl From<&str> for TaskStatus {
+        fn from(str: &str) -> TaskStatus {
+            match str {
+                "RUNNING" => TaskStatus::RUNNING,
+                "FINISHED::SUCCESS" => TaskStatus::FINISHED(TaskResult::SUCCESS),
+                "FINISHED::ERROR" => TaskStatus::FINISHED(TaskResult::ERROR),
+                _ => TaskStatus::NEW,
+            }
+        }
+    }
+
+    impl From<&Task> for DbTask {
+        fn from(task: &Task) -> DbTask {
+            DbTask {
+                id: task.id,
+                display_name: task.display_name.clone(),
+                status: task.status.to_string(),
+                features: task.features.join("||"),
+                flake_ref_uri: task.flake_ref.flake.clone(),
+                flake_ref_args: task.flake_ref.args.join("||"),
+            }
+        }
+    }
+
+    #[derive(Selectable, Identifiable, Queryable, Debug)]
+    #[diesel(table_name = locks)]
+    struct DbLock {
+        id: i32,
+        task_id: Uuid,
+        name: String,
+        type_: String,
+    }
+
+    #[derive(Insertable, Debug)]
+    #[diesel(table_name = locks)]
+    struct NewDbLock {
+        task_id: Uuid,
+        name: String,
+        type_: String,
+    }
+
+    impl From<DbLock> for NewDbLock {
+        fn from(value: DbLock) -> Self {
+            NewDbLock {
+                task_id: value.task_id,
+                name: value.name,
+                type_: value.type_,
+            }
+        }
+    }
+
+    impl DbLock {
+        fn from_lock(lock: &Lock, task_id: Uuid) -> Self {
+            match lock {
+                Lock::WRITE { name } => DbLock {
+                    id: -1,
+                    task_id,
+                    name: name.clone(),
+                    type_: "WRITE".to_string(),
+                },
+                Lock::READ { name } => DbLock {
+                    id: -1,
```

Comment:
This seems like a bad idea to me? 
If this is needed, we might want to have a clarifing comment here.

Reply:
It sucks, I know. But it works. Maybe if we fix the whole type transitioning issue for diesel, then we could get rid of this. Comment is added though